cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -397,14 +400,198 @@ struct BoxScalar<Decimal256Type> {
   static void Box(T val, Scalar* out) { checked_cast<ScalarType*>(out)->value = val; }
 };
 
-// A VisitArrayDataInline variant that calls its visitor function with logical
+// ----------------------------------------------------------------------
+// Like VisitArrayDataInline, but for ArraySpans
```

Comment:
TODO: move this to visit_data_inline.h

Question for the reviewers: do you think it's worth maintaining the VisitArrayDataInline code path if we can migrate everything to use ArraySpans? There's a lot of other code in the codebase that could probably be simplified and also probably even be faster to compile

Reply:
If we can migrate everything I would rather have only one code path.

Reply:
done (removed VisitArrayDataInline)! I added implicit construction of ArraySpan which indeed feels more natural and makes the "unboxing" to raw pointers happen in a single place, rather than having code generated that duplicates the unboxing each time this template is instantiated.